Source/Core/Core/IOS/USB/Emulated/Microphone.h

```diff
@@ -0,0 +1,131 @@
+#include <type_traits>
+
+#include "Common/CommonTypes.h"
+
+#ifdef HAVE_CUBEB
+#include "AudioCommon/CubebUtils.h"
+
+struct cubeb;
+struct cubeb_stream;
+#endif
+
+namespace IOS::HLE::USB
+{
+struct WiiSpeakState;
+
+class Microphone final
+{
+public:
+  using FloatType = float;
+
+  Microphone(const WiiSpeakState& sampler);
+  ~Microphone();
+
+  bool HasData(u32 sample_count) const;
+  u16 ReadIntoBuffer(u8* ptr, u32 size);
+  u16 GetLoudnessLevel() const;
+  void UpdateLoudness(std::ranges::input_range auto&& samples);
+  const WiiSpeakState& GetSampler() const;
+  FloatType ComputeGain(FloatType relative_db) const;
+  void SetSamplingRate(u32 sampling_rate);
+
+private:
+#ifdef HAVE_CUBEB
+  static long DataCallback(cubeb_stream* stream, void* user_data, const void* input_buffer,
+                           void* output_buffer, long nframes);
+#endif
+
+  void StreamInit();
+  void StreamTerminate();
+  void StreamStart(u32 sampling_rate);
+  void StreamStop();
+
+  using SampleType = s16;
+  static constexpr u32 BUFF_SIZE_SAMPLES = 32;
+  static constexpr u32 STREAM_SIZE = BUFF_SIZE_SAMPLES * 500;
+
+  std::array<SampleType, STREAM_SIZE> m_stream_buffer{};
+  u32 m_stream_wpos = 0;
+  u32 m_stream_rpos = 0;
+  u32 m_samples_avail = 0;
+
+  // TODO: Find how this level is calculated on real hardware
+  u16 m_loudness_level = 0;
+  struct Loudness
+  {
+    using SampleType = s16;
+    using UnsignedSampleType = std::make_unsigned_t<SampleType>;
+
+    void Update(const auto& samples)
+    {
+      samples_count += static_cast<u16>(samples.size());
+
+      const auto [min_element, max_element] = std::ranges::minmax_element(samples);
+      peak_min = std::min<SampleType>(*min_element, peak_min);
+      peak_max = std::max<SampleType>(*max_element, peak_max);
+
+      const auto begin = samples.begin();
+      const auto end = samples.end();
+      absolute_sum = std::reduce(begin, end, absolute_sum,
+                                 [](u32 a, SampleType b) { return a + std::abs(b); });
+      square_sum = std::reduce(begin, end, square_sum, [](FloatType a, s16 b) {
+        return a + std::pow(FloatType(b), FloatType(2));
+      });
+    }
+
+    SampleType GetPeak() const;
+    FloatType GetDecibel(FloatType value) const;
+    FloatType GetAmplitude() const;
+    FloatType GetAmplitudeDb() const;
+    FloatType GetAbsoluteMean() const;
+    FloatType GetAbsoluteMeanDb() const;
+    FloatType GetRootMeanSquare() const;
+    FloatType GetRootMeanSquareDb() const;
+    FloatType GetCrestFactor() const;
+    FloatType GetCrestFactorDb() const;
+    FloatType ComputeGain(FloatType db) const;
+
+    void Reset();
+    void LogStats();
+
+    // Samples used to compute the loudness level (arbitrarily chosen)
+    static constexpr u16 SAMPLES_NEEDED = 128;
+    static_assert((SAMPLES_NEEDED % BUFF_SIZE_SAMPLES) == 0);
+
+    static constexpr FloatType MAX_AMPLITUDE =
+        FloatType{std::numeric_limits<UnsignedSampleType>::max() / 2};
```

Comment:
This doesn't resolve the warning. You're still doing integer division `(65535 / 2)`. Do you want the value to be `32767.5f` or `32767` ?

Reply:
I want an integer division (and its truncation) `32767` to be converted to a float type (here float but could be changed later).

Where can I find the warning?

Reply:
The warning comes from this: https://clang.llvm.org/extra/clang-tidy/checks/bugprone/integer-division.html

Reply:
I had to use an integer cast instead, in order to silence the warning.